prov/rxm/src/rxm_ep.c

```diff
@@ -1319,6 +1390,109 @@ rxm_ep_send_common(struct rxm_ep *rxm_ep, const struct iovec *iov, void **desc,
 	}
 }
 
+static void
+rxm_ep_conn_progress_deferred_queue(struct rxm_ep *rxm_ep,
+				    struct rxm_conn *rxm_conn)
+{
+	struct rxm_tx_entry *tx_entry;
+	ssize_t ret = 0;
+
+	while (!dlist_empty(&rxm_conn->deferred_tx_queue) && !ret) {
+		tx_entry = container_of(rxm_conn->deferred_tx_queue.next,
+					struct rxm_tx_entry, deferred_tx_entry);
+		switch (tx_entry->state) {
+		case RXM_TX:		/* Eager */
+		case RXM_LMT_ACK_SENT:	/* RNDV (LMT TX ack) */
+			ret = rxm_ep_normal_send(rxm_ep, rxm_conn, tx_entry,
+						 tx_entry->deferred_pkt_size);
+			if (OFI_UNLIKELY(ret)) {
+				/* TODO: if (ret != -FI_EAGAIN), write CQ err entry */
+				assert(ret == -FI_EAGAIN);
+				break;
+			}
+			rxm_ep_dequeue_deferred_tx_queue(tx_entry);
+			break;
+		case RXM_LMT_TX:	/* RNDV (LMT TX req) */
+			if (tx_entry->deferred_pkt_size) {
+				ret = rxm_ep_normal_send(rxm_ep, rxm_conn, tx_entry,
+							 tx_entry->deferred_pkt_size);
+				if (OFI_UNLIKELY(ret)) {
+					/* TODO: if (ret != -FI_EAGAIN), write CQ err entry */
+					assert(ret == -FI_EAGAIN);
+					break;
+				}
+				tx_entry->deferred_pkt_size = 0;
+			}
+			/* RxM must keep the TX entry in the deferred TX queue until
+			 * LMT ACK is received */
+			/* fall through */
+		case RXM_LMT_ACK_WAIT:
+		case RXM_LMT_ACK_RECVD:
+			ret = -FI_EAGAIN;
+			break;
+		case RXM_LMT_READ:
+		    ret = fi_readv(tx_entry->conn->msg_ep,
+				   tx_entry->rma_buf->rxm_iov.iov,
+				   tx_entry->rma_buf->rxm_iov.desc,
+				   tx_entry->rma_buf->rxm_iov.count, 0,
+				   tx_entry->rma_buf->rxm_rma_iov.iov[0].addr,
+				   tx_entry->rma_buf->rxm_rma_iov.iov[0].key,
+				   tx_entry->rx_buf);
+		    if (!tx_entry->rma_buf->rxm_iov.count) {
+			ret = -FI_EAGAIN;
+			tx_entry->rma_buf->rxm_iov.count = 1;
+		    }   else {
+			    ret = 0;
+		    }
+			if (OFI_UNLIKELY(ret)) {
+				/* TODO: if (ret != -FI_EAGAIN), write CQ err entry */
+				assert(ret == -FI_EAGAIN);
+				break;
+			}
+			rxm_ep_dequeue_deferred_tx_queue(tx_entry);
+			free(tx_entry->rma_buf);
+			free(tx_entry);
+			break;
+		case RXM_SAR_TX:	/* SAR (TX segments) */
+			while (!dlist_empty(&tx_entry->deferred_tx_buf_list) && !ret) {
+				struct rxm_tx_buf *tx_buf =
+					container_of(tx_entry->deferred_tx_buf_list.next,
+						     struct rxm_tx_buf, hdr.entry);
+				ret = fi_send(rxm_conn->msg_ep, &tx_buf->pkt,
+					      sizeof(struct rxm_pkt) + tx_buf->pkt.ctrl_hdr.seg_size,
+					      tx_buf->hdr.desc, 0, tx_buf);
+				if (OFI_UNLIKELY(ret)) {
+					/* TODO: if (ret != -FI_EAGAIN), write CQ err entry*/
+					assert(ret == -FI_EAGAIN);
+					break;
+				}
+				dlist_remove(&tx_buf->hdr.entry);
+			}
+			if (dlist_empty(&tx_entry->deferred_tx_buf_list))
+				rxm_ep_dequeue_deferred_tx_queue(tx_entry);
+			break;
```

Comment:
The above code is same as the one in handling SAR segment completion. Can this be wrapped in a function?

Reply:
Good catch. This code has been wrapped in the function to use in both pieces of code.